crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -192,29 +220,42 @@ impl<'db> NarrowingConstraintsBuilder<'db> {
             {
                 // SAFETY: we should always have a symbol for every Name node.
                 let symbol = self.symbols().symbol_id_by_name(id).unwrap();
-                let comp_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
-                match op {
-                    ast::CmpOp::IsNot => {
-                        if comp_ty.is_singleton() {
+                let rhs_ty = inference.expression_ty(right.scoped_ast_id(self.db, scope));
+                match (op, negative) {
```

Comment:
We could make the match arms a bit less repetitive/verbose here (though arguably also less explicit?) if we call a function here that returns the negated `ast::CmpOp` if `negative` and then match on the resulting `CmpOp`. What do you think?